### PR TITLE
Add support for swapping palettes

### PR DESCRIPTION
**Pull request by mgasil**
Basic support for swapping palettes.
Example: `RenderICNSprite( ICN::PEASANT, 1, PAL::GRAY )`
Default palette is set to PAL::STANDARD.

relates to #408

### CODE REVIEW

File: src/fheroes2/agg/agg.cpp [hunk 1]
```diff
@@ -1921,20 +1916,7 @@ bool AGG::Init( void )
...
-        SDL_Color cols;
-
-        cols.r = kb_pal[index] << 2;
-        cols.g = kb_pal[index + 1] << 2;
-        cols.b = kb_pal[index + 2] << 2;
-
-        pal_colors.push_back( cols );
-    }
-    Surface::SetDefaultPalette( &pal_colors[0], pal_colors.size() );
+   PAL::InitAllPalettes();
```
Comment (shprotru) [inline src/fheroes2/agg/agg.cpp:1919]: 4 spaces instead of 3

**Comment by ihhub** [comment]
Hi @mgasil , thank you very much for you efforts! Could you please explain the intention of doing this?

**Comment by ihhub** [comment]
Hi @idshibanov , are you familiar with these tables in the original game? Any ideas where they can be useful and where they came from?

**Comment by idshibanov** [comment]
> Hi @idshibanov , are you familiar with these tables in the original game? Any ideas where they can be useful and where they came from?

@ihhub I've seen that OG game draws different effects by palette copy, direct update & swap in context of one sprite. Examples are petrfiy and bloodlust effects - it slowly redraws the unit sprite with fixed delay (75ms or 110ms) updating the palette each step. I also know that CCYCLE00-08.BIN files are used in that process somehow, but it's hard to understand.

**Comment by mgasil** [comment]
Hi @ihhub, how the tables can be used, can be found here:  [HEROES2W.c](https://github.com/jkoppel/project-ironfist/blob/master/src/raw_decompiled/HEROES2W.c)

**Comment by ihhub** [comment]
> Hi @ihhub, how the tables can be used, can be found here: [HEROES2W.c](https://github.com/jkoppel/project-ironfist/blob/master/src/raw_decompiled/HEROES2W.c)

Hi @mgasil , thanks for this!

**Comment by idshibanov** [comment]
I like this, it is cleaner. The actual usage of palettes in different contexts require more testing/discussion though because we currently apply palette at game load rather than a draw call.

**Comment by ihhub** [comment]
Hi @mgasil , thank you very much for the contribution to the project!

**Comment by ihhub** [comment]
Hi @mgasil , could you please share the information where did you get these tables? We are missing some tables in relation to shadowing and air elemental and other small effects and these tables would be extremely helpful for the project.

**Pull request merged**